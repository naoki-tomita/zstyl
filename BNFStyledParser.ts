type Types =
    "StyleSheet"
  | "LocalStyle"
  | "Style"
  | "Identifier"
  | "Value"
  | "Values"
  | "NestedStyle"
  | "Selector"
  | "Block"
  | "MediaCondition"
  | "Media"
  | "Keyframe"
  ;

interface Ast {
  type: Types;
};

type RootStyle = LocalStyle | NestedStyle | MediaStyle | KeyframeStyle;

export interface StyleSheetAst extends Ast {
  type: "StyleSheet";
  children: Array<LocalStyle | NestedStyle | MediaStyle | KeyframeStyle>;
}

export const StyleSheetParser: Parser<StyleSheetAst> = {
  parse(style: string): ParsedResult<StyleSheetAst> {
    const children: StyleSheetAst["children"] = [];
    let result: ParsedResult<RootStyle> = { remaining: style };
    let successToParse;
    do {
      successToParse = false;
      for (const parser of [LocalStyleParser, NestedStyleParser, MediaStyleParser, KeyframeStyleParser]) {
        result = parser.parse(result.remaining);
        if (result.ast) {
          children.push(result.ast);
          successToParse = true;
        }
      }
    } while (successToParse)

    return {
      ast: {
        type: "StyleSheet",
        children,
      },
      remaining: result.remaining
    };
  }
}

export interface Identifier extends Ast {
  type: "Identifier";
  name: string;
}

export const IdentifierParser: Parser<Identifier> = {
  parse(style: string): ParsedResult<Identifier> {
    style = style.trimStart();
    const result = style.match(/^([a-zA-Z0-9\-_.]+)/);
    if (result == null || result[0].length === 0) {
      return { remaining: style };
    }
    const name = result[0];
    return {
      ast: {
        type: "Identifier",
        name,
      },
      remaining: style.slice(name.length),
    }
  }
}

export interface Value extends Ast {
  type: "Value";
  value: string;
}

export const ValueParser: Parser<Value> = {
  parse(style: string): ParsedResult<Value> {
    const result = style.trimStart().match(/^([a-zA-Z0-9\-%#_.\()]+)/);
    if (result == null || result[0].length === 0) {
      return { remaining: style };
    }
    const value = result[0];
    return {
      ast: {
        type: "Value",
        value,
      },
      remaining: style.trimStart().slice(value.length),
    }
  }
}

export interface Values extends Ast {
  type: "Values";
  values: Value[];
}

export const ValuesParser: Parser<Values> = {
  parse(style: string): ParsedResult<Values> {
    let result = ValueParser.parse(style);
    if (!result.ast) {
      return { remaining: style };
    }
    const results: Value[] = [result.ast];
    while (true) {
      const remaining = result.remaining;
      result = ValueParser.parse(remaining);
      if (!result.ast) {
        return {
          ast: {
            type: "Values",
            values: results,
          },
          remaining,
        }
      }
      results.push(result.ast);
    }
  }
}

export interface Style extends Ast {
  type: "Style"
  prop: Identifier;
  values: Values;
}
export const StyleParser: Parser<Style> = {
  parse(style: string): ParsedResult<Style> {
    const identifier = IdentifierParser.parse(style);
    if (!identifier.ast) {
      return { remaining: style };
    }
    const colonWithIdentifiers = identifier.remaining.trimStart();
    if (!colonWithIdentifiers.startsWith(":")) {
      return { remaining: style };
    }
    const identifiers = ValuesParser.parse(colonWithIdentifiers.slice(1));
    if (!identifiers.ast) {
      return { remaining: style };
    }
    return {
      ast: {
        type: "Style",
        prop: identifier.ast,
        values: identifiers.ast,
      },
      remaining: identifiers.remaining
    }
  }
}

export interface LocalStyle extends Ast {
  type: "LocalStyle";
  prop: Identifier;
  values: Values;
}
export const LocalStyleParser: Parser<LocalStyle> = {
  parse(style: string): ParsedResult<LocalStyle> {
    const result = StyleParser.parse(style);
    if (
      !result.ast ||
      !result.remaining.trimStart().startsWith(";")
    ) {
      return { remaining: style };
    }
    return {
      ast: {
        type: "LocalStyle",
        prop: result.ast.prop,
        values: result.ast.values,
      },
      remaining: result.remaining.trimStart().slice(1)
    }
  }
}

type AnyStyle = LocalStyle | NestedStyle;

export interface Selector extends Ast {
  type: "Selector";
  value: string; // ここを BNF で表現するのはあきらめました
}

export const SelectorParser: Parser<Selector> = {
  parse(style: string): ParsedResult<Selector> {
    const result = style.trimStart().match(/^([a-zA-Z0-9_()+>| ,.#~=%^$&\[\]"'*:/\-]+)/)
    if (result == null) {
      return { remaining: style };
    }
    const value = result[0].trim()
    return {
      ast: {
        type: "Selector",
        value,
      },
      remaining: style.trimStart().slice(value.length)
    }
  }
}

export interface Block extends Ast {
  type: "Block";
  body: AnyStyle[];
}

export const BlockParser: Parser<Block> = {
  parse(style: string): ParsedResult<Block> {
    if (!style.trimStart().startsWith("{")) {
      return { remaining: style };
    }
    let result: ParsedResult<AnyStyle> = { remaining: style.trimStart().slice(1) };
    const results: AnyStyle[] = [];
    let successToParse: boolean;
    do {
      successToParse = false;
      for (const parser of [LocalStyleParser, NestedStyleParser]) {
        result = parser.parse(result.remaining);
        if (result.ast) {
          successToParse = true;
          results.push(result.ast);
        }
      }
    } while (successToParse)
    if (!result.remaining.trimStart().startsWith("}")) {
      return { remaining: style }
    }
    return {
      ast: {
        type: "Block",
        body: results,
      },
      remaining: result.remaining.trimStart().slice(1)
    }
  }
}

export interface NestedStyle extends Ast {
  type: "NestedStyle";
  selector: Selector;
  block: Block;
}

export const NestedStyleParser: Parser<NestedStyle> = {
  parse(style: string): ParsedResult<NestedStyle> {
    const selector = SelectorParser.parse(style);
    if (!selector.ast) {
      return { remaining: style };
    }
    const block = BlockParser.parse(selector.remaining);
    if (!block.ast) {
      return { remaining: style };
    }
    return {
      ast: {
        type: "NestedStyle",
        selector: selector.ast,
        block: block.ast,
      },
      remaining: block.remaining
    }
  }
}

interface MediaCondition extends Ast {
  type: "MediaCondition";
  name: Identifier;
  value: Identifier;
}

export interface MediaStyle extends Ast {
  type: "Media"
  condition: MediaCondition;
  block: Block;
}

export const MediaStyleParser: Parser<MediaStyle> = {
  parse(style: string): ParsedResult<MediaStyle> {
    if (!style.trimStart().startsWith("@media")) {
      return { remaining: style };
    }
    const slicedAtMediaStyle = style.trimStart().slice("@media".length);
    if (!slicedAtMediaStyle.trimStart().startsWith("(")) {
      return { remaining: style };
    }
    const leftParenTrimmedStyle = slicedAtMediaStyle.trimStart().slice(1);
    const nameResult = IdentifierParser.parse(leftParenTrimmedStyle);
    if (!nameResult.ast) {
      return { remaining: style };
    }
    if (!nameResult.remaining.trimStart().startsWith(":")) {
      return { remaining: style };
    }
    const valueResult = IdentifierParser.parse(nameResult.remaining.trimStart().slice(1));
    if (!valueResult.ast) {
      return { remaining: style };
    }
    if (!valueResult.remaining.trimStart().startsWith(")")) {
      return { remaining: style };
    }
    const rightParenTrimmedStyle = valueResult.remaining.trimStart().slice(1);
    const blockResult = BlockParser.parse(rightParenTrimmedStyle);
    if (!blockResult.ast) {
      return { remaining: style };
    }
    return {
      ast: {
        type: "Media",
        condition: {
          type: "MediaCondition",
          name: nameResult.ast,
          value: valueResult.ast,
        },
        block: blockResult.ast
      },
      remaining: blockResult.remaining
    }
  }
}

export interface KeyframeStyle extends Ast {
  type: "Keyframe"
  name: Identifier;
  block: Block;
}

export const KeyframeStyleParser: Parser<KeyframeStyle> = {
  parse(style: string): ParsedResult<KeyframeStyle> {
    if (!style.trimStart().startsWith("@keyframe")) {
      return { remaining: style };
    }
    const keyframeExcludedStyle = style.trimStart().slice("@keyframe".length);
    const nameResult = IdentifierParser.parse(keyframeExcludedStyle);
    if (!nameResult.ast) {
      return { remaining: style };
    }
    const blockStyle = nameResult.remaining;
    const blockResult = BlockParser.parse(blockStyle);
    if (!blockResult.ast) {
      return { remaining: style };
    }
    return {
      ast: {
        type: "Keyframe",
        name: nameResult.ast,
        block: blockResult.ast
      },
      remaining: blockResult.remaining,
    }
  }
}

type ParsedResult<T extends Ast> = {
  ast?: T, remaining: string
};

interface Parser<T extends Ast> {
  parse(style: string): ParsedResult<T>;
}
