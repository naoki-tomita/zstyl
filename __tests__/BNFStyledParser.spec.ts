import { BlockParser, IdentifierParser, ValuesParser, KeyframeStyleParser, LocalStyleParser, MediaStyleParser, NestedStyleParser, SelectorParser, StyleParser, StyleSheetParser, ValueParser } from "../BNFStyledParser";
describe("IdentifierParser", () => {
  const tests = [
    {
      input: "",
      expected: { remaining: "" },
    },
    {
      input: "foo",
      expected: {
        ast: {
          type: "Identifier",
          name: "foo",
        },
        remaining: ""
      },
    },
    {
      input: "foo bar",
      expected: {
        ast: {
          type: "Identifier",
          name: "foo",
        },
        remaining: " bar"
      },
    },
    {
      input: ",foo bar",
      expected: {
        remaining: ",foo bar"
      },
    },
    {
      input: " foo bar ",
      expected: {
        ast: {
          type: "Identifier",
          name: "foo"
        },
        remaining: " bar "
      }
    }
  ]

  tests.forEach(({ input, expected }) => {
    it(`should be parsed. "${input}"`, () => {
      const actual = IdentifierParser.parse(input)
      expect(actual).toEqual(expected);
    });
  });
});

describe("ValueParser", () => {
  const tests = [
    {
      input: "",
      expected: { remaining: "" },
    },
    {
      input: "foo",
      expected: {
        ast: { type: "Value", value: "foo" },
        remaining: ""
      },
    },
    {
      input: "foo bar",
      expected: {
        ast: { type: "Value", value: "foo" },
        remaining: " bar"
      },
    },
    {
      input: ",foo bar",
      expected: {
        remaining: ",foo bar"
      },
    },
    {
      input: " foo bar ",
      expected: {
        ast: { type: "Value", value: "foo" },
        remaining: " bar "
      }
    },
    {
      input: "#ccc",
      expected: {
        ast: { type: "Value", value: "#ccc" },
        remaining: ""
      }
    },
    {
      input: "3.25em",
      expected: {
        ast: { type: "Value", value: "3.25em" },
        remaining: ""
      }
    },
    {
      input: "var(--nc-tx-1)",
      expected: {
        ast: { type: "Value", value: "var(--nc-tx-1)" },
        remaining: ""
      }
    }
  ]

  tests.forEach(({ input, expected }) => {
    it(`should be parsed. "${input}"`, () => {
      const actual = ValueParser.parse(input)
      expect(actual).toEqual(expected);
    });
  });
});

describe("ValuesParser", () => {
  const tests = [
    {
      input: "",
      expected: { remaining: "" },
    },
    {
      input: "foo",
      expected: {
        ast: {
          type: "Values",
          values: [
            { type: "Value", value: "foo" }
          ]
        },
        remaining: ""
      },
    },
    {
      input: "foo bar",
      expected: {
        ast: {
          type: "Values",
          values: [
            { type: "Value", value: "foo" },
            { type: "Value", value: "bar" }
          ]
        },
        remaining: ""
      },
    },
    {
      input: ",foo bar",
      expected: {
        remaining: ",foo bar"
      },
    },
    {
      input: " foo bar ",
      expected: {
        ast: {
          type: "Values",
          values: [
            { type: "Value", value: "foo" },
            { type: "Value", value: "bar" }
          ]
        },
        remaining: " "
      }
    },
    {
      input: " foo bar;",
      expected: {
        ast: {
          type: "Values",
          values: [
            { type: "Value", value: "foo" },
            { type: "Value", value: "bar" }
          ]
        },
        remaining: ";"
      }
    }
  ]

  tests.forEach(({ input, expected }) => {
    it(`should be parsed. "${input}"`, () => {
      const actual = ValuesParser.parse(input)
      expect(actual).toEqual(expected);
    });
  });
});

describe("StyleParser", () => {
  const tests = [
    {
      input: "",
      expected: { remaining: "" },
    },
    {
      // identifier
      input: "foo",
      expected: {
        remaining: "foo"
      },
    },
    {
      // identifier with colon
      input: "foo:",
      expected: {
        remaining: "foo:"
      },
    },
    {
      // identifier with colon and semicolon
      input: "foo: ;",
      expected: {
        remaining: "foo: ;"
      },
    },
    {
      // valid Style
      input: "foo: bar",
      expected: {
        ast: {
          type: "Style",
          prop: {
            type: "Identifier",
            name: "foo"
          },
          values: {
            type: "Values",
            values: [
              { type: "Value", value: "bar" },
            ]
          }
        },
        remaining: ""
      }
    },
    {
      input: " foo: bar hoge ;",
      expected: {
        ast: {
          type: "Style",
          prop: {
            type: "Identifier",
            name: "foo"
          },
          values: {
            type: "Values",
            values: [
              { type: "Value", value: "bar" },
              { type: "Value", value: "hoge" },
            ]
          }
        },
        remaining: " ;"
      }
    }
  ]

  tests.forEach(({ input, expected }) => {
    it(`should be parsed. "${input}"`, () => {
      const actual = StyleParser.parse(input)
      expect(actual).toEqual(expected);
    });
  });
});

describe("LocalStyleParser", () => {
  const tests = [
    {
      input: " foo: bar hoge ; foo: bar;",
      expected: {
        ast: {
          type: "LocalStyle",
          prop: {
            type: "Identifier",
            name: "foo"
          },
          values: {
            type: "Values",
            values: [
              { type: "Value", value: "bar" },
              { type: "Value", value: "hoge" },
            ]
          }
        },
        remaining: " foo: bar;"
      }
    }
  ]

  tests.forEach(({ input, expected }) => {
    it(`should be parsed. "${input}"`, () => {
      const actual = LocalStyleParser.parse(input)
      expect(actual).toEqual(expected);
    });
  });
});

describe("SelectorParser", () => {
  const tests = [
    {
      input: "",
      expected: {
        remaining: ""
      },
    },
    {
      input: "hoge",
      expected: {
        ast: {
          type: "Selector",
          value: "hoge",
        },
        remaining: ""
      }
    },
    {
      input: "hoge { ... }",
      expected: {
        ast: {
          type: "Selector",
          value: "hoge",
        },
        remaining: " { ... }"
      }
    },
    {
      input: "#hoge",
      expected: {
        ast: {
          type: "Selector",
          value: "#hoge",
        },
        remaining: ""
      }
    },
    {
      input: ".hoge",
      expected: {
        ast: {
          type: "Selector",
          value: ".hoge",
        },
        remaining: ""
      }
    },
    {
      input: "hoge::hover",
      expected: {
        ast: {
          type: "Selector",
          value: "hoge::hover",
        },
        remaining: ""
      }
    },
    {
      input: "hoge::nth-child(1)",
      expected: {
        ast: {
          type: "Selector",
          value: "hoge::nth-child(1)",
        },
        remaining: ""
      }
    },
    {
      input: `a[title] {`,
      expected: {
        ast: {
          type: "Selector",
          value: "a[title]",
        },
        remaining: " {"
      }
    },
    {
      input: `a[href="https://example.org"] {`,
      expected: {
        ast: {
          type: "Selector",
          value: `a[href="https://example.org"]`,
        },
        remaining: " {"
      }
    },
    {
      input: `a[href*="example"] {`,
      expected: {
        ast: {
          type: "Selector",
          value: `a[href*="example"]`,
        },
        remaining: " {"
      }
    },
    {
      input: `a[href$=".org"] {`,
      expected: {
        ast: {
          type: "Selector",
          value: `a[href$=".org"]`,
        },
        remaining: " {"
      }
    },
    {
      input: `a[class~="logo"] {`,
      expected: {
        ast: {
          type: "Selector",
          value: `a[class~="logo"]`,
        },
        remaining: " {"
      }
    },
    {
      input: "&::hover",
      expected: {
        ast: {
          type: "Selector",
          value: `&::hover`,
        },
        remaining: ""
      }
    }
  ]

  tests.forEach(({ input, expected }) => {
    it(`should be parsed. ${input}`, () => {
      const actual = SelectorParser.parse(input);
      expect(actual).toEqual(expected);
    });
  })
});

describe("BlockParser", () => {
  const tests = [
    {
      input: "hoge { fuga: foo; }",
      expected: {
        remaining: "hoge { fuga: foo; }",
      }
    },
    {
      input: "{ fuga: foo; } hoge",
      expected: {
        ast: {
          type: "Block",
          body: [
            {
              type: "LocalStyle",
              prop: {
                type: "Identifier",
                name: "fuga"
              },
              values: {
                type: "Values",
                values: [
                  { type: "Value", value: "foo" }
                ]
              }
            },
          ]
        },
        remaining: " hoge",
      }
    },
    {
      input: "{ fuga: foo;",
      expected: {
        remaining: "{ fuga: foo;",
      }
    },
    {
      input: "{}",
      expected: {
        ast: {
          type: "Block",
          body: []
        },
        remaining: "",
      }
    }
  ]

  tests.forEach(({ input, expected }) => {
    it(`should be parsed.  ${input}`, () => {
      const actual = BlockParser.parse(input);
      expect(actual).toEqual(expected);
    });
  })
});

describe("NestedStyleParser", () => {
  const tests = [
    {
      input: "hoge { fuga: foo; }",
      expected: {
        ast: {
          type: "NestedStyle",
          selector: {
            type: "Selector",
            value: "hoge"
          },
          block: {
            type: "Block",
            body: [
              {
                type: "LocalStyle",
                prop: {
                  type: "Identifier",
                  name: "fuga"
                },
                values: {
                  type: "Values",
                  values: [
                    { type: "Value", value: "foo" }
                  ]
                }
              }
            ]
          }
        },
        remaining: "",
      }
    },
    {
      // multiple nested style.
      input: `
hoge {
  fuga: foo;

  foo {
    bar: piyo;
  }
}
`.trim(),
      expected: {
        ast: {
          type: "NestedStyle",
          selector: {
            type: "Selector",
            value: "hoge"
          },
          block: {
            type: "Block",
            body: [
              {
                type: "LocalStyle",
                prop: {
                  type: "Identifier",
                  name: "fuga"
                },
                values: {
                  type: "Values",
                  values: [
                    { type: "Value", value: "foo" },
                  ]
                }
              },
              {
                type: "NestedStyle",
                selector: {
                  type: "Selector",
                  value: "foo"
                },
                block: {
                  type: "Block",
                  body: [
                    {
                      type: "LocalStyle",
                      prop: {
                        type: "Identifier",
                        name: "bar"
                      },
                      values: {
                        type: "Values",
                        values: [
                          { type: "Value", value: "piyo" }
                        ]
                      }
                    }
                  ]
                }
              }
            ]
          }
        },
        remaining: "",
      }
    }
  ]

  tests.forEach(({ input, expected }) => {
    it("should be parsed. ", () => {
      const actual = NestedStyleParser.parse(input);
      expect(actual).toEqual(expected);
    });
  })
});

describe("MediaStyleParser", () => {
  const tests = [
    {
      input: "@media(min-width:120px){}",
      expected: {
        ast: {
          type: "Media",
          condition: {
            type: "MediaCondition",
            name: {
              type: "Identifier",
              name: "min-width"
            },
            value: {
              type: "Identifier",
              name: "120px",
            }
          },
          block: {
            type: "Block",
            body: []
          }
        },
        remaining: ""
      }
    },
    {
      input: " @media  ( min-width: 120px )  {}",
      expected: {
        ast: {
          type: "Media",
          condition: {
            type: "MediaCondition",
            name: {
              type: "Identifier",
              name: "min-width"
            },
            value: {
              type: "Identifier",
              name: "120px",
            }
          },
          block: {
            type: "Block",
            body: []
          }
        },
        remaining: ""
      }
    },
    {
      input: "@media foo {",
      expected: {
        remaining: "@media foo {"
      }
    },
    {
      input: "@media foo a",
      expected: {
        remaining: "@media foo a"
      }
    },
    {
      input: "@media (min-width: 120px hoge",
      expected: {
        remaining: "@media (min-width: 120px hoge"
      }
    },
    {
      input: "@media (",
      expected: {
        remaining: "@media ("
      }
    },
    {
      input: "@media",
      expected: {
        remaining: "@media"
      }
    }
  ]
  tests.forEach(({ input, expected }) => {
    it(`should be parse "${input}"`, () => {
      const actual = MediaStyleParser.parse(input);
      expect(actual).toEqual(expected);
    });
  });
});

describe("KeyframeStyleParser", () => {
  const tests = [
    {
      input: "@keyframes foo {}",
      expected: {
        ast: {
          type: "Keyframe",
          name: {
            type: "Identifier",
            name: "foo"
          },
          block: {
            type: "Block",
            body: []
          }
        },
        remaining: ""
      }
    },
    {
      input: " @keyframes  foo  {}",
      expected: {
        ast: {
          type: "Keyframe",
          name: {
            type: "Identifier",
            name: "foo"
          },
          block: {
            type: "Block",
            body: []
          }
        },
        remaining: ""
      }
    },
    {
      input: "@keyframes foo {",
      expected: {
        remaining: "@keyframes foo {"
      }
    },
    {
      input: "@keyframes foo a",
      expected: {
        remaining: "@keyframes foo a"
      }
    },
    {
      input: "@keyframes foo hoge",
      expected: {
        remaining: "@keyframes foo hoge"
      }
    },
    {
      input: "@keyframe",
      expected: {
        remaining: "@keyframe"
      }
    },
    {
      input: "@keyframe",
      expected: {
        remaining: "@keyframe"
      }
    },
  ]
  tests.forEach(({ input, expected }) => {
    it(`should be parse "${input}"`, () => {
      const actual = KeyframeStyleParser.parse(input);
      expect(actual).toEqual(expected);
    });
  });
});

describe("StyleSheetParser", () => {
  it("should parse stylesheet", () => {
    const style = `
display: flex;

&:hover {
  color: red;

  .inner {
    color: black;
  }
}

.wrapper {
  color: blue;

  .inner {
    color: green;
  }
}

@media (max-width: 200px) {
  display: inline-flex;

  .inner {
    color: yellow;
  }
}

@keyframes colors {
  from {
    color: yellow;
  }

  50% {
    color: black;
  }

  to {
    color: red;
  }
}
    `;
    const actual = StyleSheetParser.parse(style);
    expect(actual).toMatchSnapshot();
  });
});
