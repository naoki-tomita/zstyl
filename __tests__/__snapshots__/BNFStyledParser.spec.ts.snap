// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StyleSheetParser should parse stylesheet 1`] = `
{
  "ast": {
    "children": [
      {
        "prop": {
          "name": "display",
          "type": "Identifier",
        },
        "type": "LocalStyle",
        "values": {
          "type": "Identifiers",
          "values": [
            {
              "name": "flex",
              "type": "Identifier",
            },
          ],
        },
      },
      {
        "block": {
          "body": [
            {
              "prop": {
                "name": "color",
                "type": "Identifier",
              },
              "type": "LocalStyle",
              "values": {
                "type": "Identifiers",
                "values": [
                  {
                    "name": "red",
                    "type": "Identifier",
                  },
                ],
              },
            },
            {
              "block": {
                "body": [
                  {
                    "prop": {
                      "name": "color",
                      "type": "Identifier",
                    },
                    "type": "LocalStyle",
                    "values": {
                      "type": "Identifiers",
                      "values": [
                        {
                          "name": "black",
                          "type": "Identifier",
                        },
                      ],
                    },
                  },
                ],
                "type": "Block",
              },
              "selector": {
                "type": "Selector",
                "value": ".inner",
              },
              "type": "NestedStyle",
            },
          ],
          "type": "Block",
        },
        "selector": {
          "type": "Selector",
          "value": "&:hover",
        },
        "type": "NestedStyle",
      },
      {
        "block": {
          "body": [
            {
              "prop": {
                "name": "color",
                "type": "Identifier",
              },
              "type": "LocalStyle",
              "values": {
                "type": "Identifiers",
                "values": [
                  {
                    "name": "blue",
                    "type": "Identifier",
                  },
                ],
              },
            },
            {
              "block": {
                "body": [
                  {
                    "prop": {
                      "name": "color",
                      "type": "Identifier",
                    },
                    "type": "LocalStyle",
                    "values": {
                      "type": "Identifiers",
                      "values": [
                        {
                          "name": "green",
                          "type": "Identifier",
                        },
                      ],
                    },
                  },
                ],
                "type": "Block",
              },
              "selector": {
                "type": "Selector",
                "value": ".inner",
              },
              "type": "NestedStyle",
            },
          ],
          "type": "Block",
        },
        "selector": {
          "type": "Selector",
          "value": ".wrapper",
        },
        "type": "NestedStyle",
      },
      {
        "block": {
          "body": [
            {
              "prop": {
                "name": "display",
                "type": "Identifier",
              },
              "type": "LocalStyle",
              "values": {
                "type": "Identifiers",
                "values": [
                  {
                    "name": "inline-flex",
                    "type": "Identifier",
                  },
                ],
              },
            },
            {
              "block": {
                "body": [
                  {
                    "prop": {
                      "name": "color",
                      "type": "Identifier",
                    },
                    "type": "LocalStyle",
                    "values": {
                      "type": "Identifiers",
                      "values": [
                        {
                          "name": "yellow",
                          "type": "Identifier",
                        },
                      ],
                    },
                  },
                ],
                "type": "Block",
              },
              "selector": {
                "type": "Selector",
                "value": ".inner",
              },
              "type": "NestedStyle",
            },
          ],
          "type": "Block",
        },
        "condition": {
          "prop": {
            "name": "max-width",
            "type": "Identifier",
          },
          "type": "Style",
          "values": {
            "type": "Identifiers",
            "values": [
              {
                "name": "200px",
                "type": "Identifier",
              },
            ],
          },
        },
        "type": "Media",
      },
      {
        "block": {
          "body": [
            {
              "block": {
                "body": [
                  {
                    "prop": {
                      "name": "color",
                      "type": "Identifier",
                    },
                    "type": "LocalStyle",
                    "values": {
                      "type": "Identifiers",
                      "values": [
                        {
                          "name": "yellow",
                          "type": "Identifier",
                        },
                      ],
                    },
                  },
                ],
                "type": "Block",
              },
              "selector": {
                "type": "Selector",
                "value": "from",
              },
              "type": "NestedStyle",
            },
            {
              "block": {
                "body": [
                  {
                    "prop": {
                      "name": "color",
                      "type": "Identifier",
                    },
                    "type": "LocalStyle",
                    "values": {
                      "type": "Identifiers",
                      "values": [
                        {
                          "name": "black",
                          "type": "Identifier",
                        },
                      ],
                    },
                  },
                ],
                "type": "Block",
              },
              "selector": {
                "type": "Selector",
                "value": "50%",
              },
              "type": "NestedStyle",
            },
            {
              "block": {
                "body": [
                  {
                    "prop": {
                      "name": "color",
                      "type": "Identifier",
                    },
                    "type": "LocalStyle",
                    "values": {
                      "type": "Identifiers",
                      "values": [
                        {
                          "name": "red",
                          "type": "Identifier",
                        },
                      ],
                    },
                  },
                ],
                "type": "Block",
              },
              "selector": {
                "type": "Selector",
                "value": "to",
              },
              "type": "NestedStyle",
            },
          ],
          "type": "Block",
        },
        "name": {
          "name": "colors",
          "type": "Identifier",
        },
        "type": "Keyframe",
      },
    ],
    "type": "StyleSheet",
  },
  "remaining": "
    ",
}
`;
